Cubeto draw

arnaudlorin opened the pull request:
Add a special structure that hold the cubes to be drawn and add and remove element as required.
This avoid to recompute the vec of cubes to be drawn at every frame.
World have an item of this type and interact with it.
For the server this struct will remain empty.
I added a special function for the server so that it do not compute the cube to be drawn

arnaudlorin committed:
*Introduce a struct holding the CubeAttr to be drawn
This vector is not recomputed at each frame as before
but instead elements are added and removed as necessary
Visibility is taken into account: neighboring cube are added/removed
from the vec
* Added a item in the debug menu for numbers of cube drawn
* Lock the mouse to the window

arnaudlorin committed:
lock the mouse to the windows

arnaudlorin committed:
Add an apply_action_no_render so that the server do no render cube

arthurBricq commented on src/cubes_to_draw.rs:
```diff
@@ -0,0 +1,117 @@
+        self.cube_to_draw.iter().position(|x| x.pos() == selected_cube.unwrap().to_cube_coordinates().as_array())
+            .map(|x| {
+                self.selected_cube_index = Some(x); 
+                self.cube_to_draw[x].set_is_selected(true as u8)});
```
Please don't use `x` as the name inside the closure, especially if in each closure the variable as a completely different 'homogenity'. First `x` could be `cube` and second `x` could be `index`.

You may think this is pointless, but because of this `x`, it's hard to understand what this code does.

arthurBricq commented on src/cubes_to_draw.rs:
```diff
@@ -0,0 +1,117 @@
+
+    pub fn buffer_to_draw(&self,display: &Display<WindowSurface>) -> glium::VertexBuffer<CubeAttr> {
+        glium::VertexBuffer::immutable(display, &self.cube_to_draw).unwrap()
+    }
```
Not used, to be removed

arthurBricq commented on src/cubes_to_draw.rs:
```diff
@@ -0,0 +1,117 @@
+        assert!(cube_to_draw.cube_to_draw().len()==1);
+
+        cube_to_draw.destroy_cube(&Vector3::newf([0.,0.,0.]));
+        assert!(cube_to_draw.cube_to_draw().len()==0);
```
It's a good practice to format your code before you submit it. This is not formatted well, there should 
- spaces before and after `==`
- spaces after `,` in list

Just use `cargo fmt` or any other formatting tool provided by your code editor

arthurBricq commented on src/graphics/cube.rs:
```diff
@@ -151,6 +151,14 @@ impl CubeAttr {
     }
+    
+    pub fn pos(&self) -> [f32;3] {
+        [self.world_matrix[3][0]-0.5,self.world_matrix[3][1]-0.5,self.world_matrix[3][2]-0.5]
```
Hm... That's not a good idea.

Why is this not good ? 
- Because it assumes an inner logic of class (`cube.rs` in the model), therefore IT BREAKS THE RESPONSABILITY PRINCIPLE. 
- The problem is that it shouldn't be `src/cube.rs` to compute the `world_matrix`. When I first created the class `Cube`, there was no distinction between model and view. So a mistake was propagated and here we are, reverse engineering the inner logic of the model in the view. 

How to fix ?
- Add a property `position: Vector3` inside `CubeAttr`
- Change the constructor of `CubeAttr` so that
  - it receives a `position` instead of the `world_matrix`
  - it computes the `world_matrix` in the constructor, using a **static method** `compute_world_matrix`
- Remove the computation of the `world_matrix` from the model

arthurBricq commented on src/graphics/cube.rs:
```diff
@@ -151,6 +151,14 @@ impl CubeAttr {
+        [self.world_matrix[3][0]-0.5,self.world_matrix[3][1]-0.5,self.world_matrix[3][2]-0.5]
+    }
+    
+    pub fn set_is_selected(&mut self, is_selected: u8) {
```
`is_selected` should be a bool. 

Again the responsability principle ahaha

It is not the responsability of other classes to know that inside `CubeAttr`, the bool is stored as a `u8` because OpenGL does not support `bool`

arthurBricq commented on src/world.rs:
```diff
@@ -168,7 +171,19 @@ impl World {
+        self.cubes_to_draw.set_selected_cube(selected_cube)
+    }
+
+    pub fn cube_to_draw(&self) -> &Vec<CubeAttr> {
```
typo, should be `cubes_to_draw`

arthurBricq commented on src/world.rs:
```diff
@@ -9,16 +9,18 @@ use crate::world_generation::perlin::PerlinNoise;
 pub struct World {
     /// The list of the chunks currently being displayed
     chunks: Vec<Chunk>,
+    cubes_to_draw: CubesToDraw,
```
This should be an `Option<CubeToDraw>`, because the server will not have this. 

The constructors of this class should create  `cubes_to_draw` with `None`

Then we add a method to instantiate `cubes_to_draw` to something, maybe named `prepare_for_opengl` ? If you have a better name, don't hesitate.

arthurBricq commented on src/world.rs:
```diff
@@ -210,8 +225,33 @@ impl World {
+        }
+    }
+
+    pub fn apply_action_no_render(&mut self, action: &Action) {
+        match action {
+            Action::Destroy { at } => {
+                self.destroy_cube(at.clone());
+
+            },
+            Action::Add { at, block } => {
+                self.add_cube(at.clone(), block.clone());
+            }
```
Remove this function. 

Again the responsability principle.

The caller must NOT KNOW ANYTHING about the inner logic of the class. 

It is the class that knows whether `cubes_to_draw` is set to `None` or not.

arthurBricq commented on src/game_server.rs:
```diff
@@ -66,7 +66,7 @@ impl GameServer {
 
     pub fn on_new_action(&mut self, client: usize, action: Action) {
-        self.world.apply_action(&action);
+        self.world.apply_action_no_render(&action);
```
This is not good. You will see my comments in `world.rs` that explains how to fix this.

arthurBricq commented on src/world_renderer.rs:
```diff
@@ -88,6 +89,13 @@ impl WorldRenderer {
+        // Try to lock the mouse to the window, this doen't work for all OS
+        let lock_mouse = window.set_cursor_grab(CursorGrabMode::Confined)
+            .or_else(|_e| window.set_cursor_grab(CursorGrabMode::Locked));
+        if lock_mouse.is_err() { println!("Can't lock")
```
formatting mistake, please let's be careful with this in the future, it's very important on large project.

arthurBricq commented on src/world_renderer.rs:
```diff
@@ -155,7 +166,7 @@ impl WorldRenderer {
                         if self.hud_renderer.show_debug() {
                             self.hud_renderer
-                                .set_debug(DebugData::new(self.fps_manager.fps(), self.cam.position().clone(), self.cam.rotation()));
+                                .set_debug(DebugData::new(self.fps_manager.fps(), self.cam.position().clone(), self.cam.rotation(),self.world.number_cubes_rendered()));
```
add a space after  `,`

arthurBricq commented on src/world_renderer.rs:
```diff
@@ -182,7 +193,8 @@ impl WorldRenderer {
                         // OpenGL instancing = instead of setting 1000 times different uniforms, you give once 1000 attributes
-                        let position_buffer = glium::VertexBuffer::dynamic(&display, &self.world.get_cubes_to_draw(self.cam.touched_cube())).unwrap();
+                        self.world.set_selected_cube(self.cam.touched_cube());
+                        let position_buffer = glium::VertexBuffer::immutable(&display, self.world.cube_to_draw()).unwrap();
```
I am assuming that there is a reason why you changed to `VertexBuffer::immutable`. I just put this comment so that you can explain me why :)

arnaudlorin replied:
Once created VertexBuffer is not modified so dynamic is not needed and I saw a small speed up of fps using immutable

arthurBricq commented on src/world.rs:
```diff
@@ -210,8 +225,33 @@ impl World {
         match action {
-            Action::Destroy { at } => self.destroy_cube(at.clone()),
-            Action::Add { at, block } => self.add_cube(at.clone(), block.clone())
+            Action::Destroy { at } => {
+                let cubes_to_add = self.destroy_cube(at.clone());
+                self.cubes_to_draw.destroy_cube(&at);
+                for cube in cubes_to_add {
+                    self.cubes_to_draw.add_cube(&cube)
+                }
+
+            },
+            Action::Add { at, block } => {
+                let (cubes_to_destroy, cube) = self.add_cube(at.clone(), block.clone());
+                self.cubes_to_draw.add_cube(&cube);
+                for position in cubes_to_destroy {
+                    self.cubes_to_draw.destroy_cube(&position)
+                }
+            }
+        }
+    }
```
After changing `cubes_to_draw` to an Option, you will have to change this function so that the extra logic is only done when the option is set to something.

Also, please add some comments that explains a little bit why we have to add cubes when deleting, as it could be confusing. The something like `// Destroying a cube can reveal other cubes to draw`. 

Actually `cube_to_add` is not a good name.
- First there is a typo, it is `cubes_to_add` 
- But I think that `revealed_cubes` is even better, because it allows to not put a comment.

arthurBricq commented on src/world.rs:
```diff
@@ -225,19 +265,30 @@ impl World {
+    fn add_cube(&mut self, at: Vector3, block: Block) -> (Vec<Vector3>,Cube) {
         // For all the neighbors positions, increase their internal counter
         let mut count = 0;
+        let mut to_destroy = Vec::new();
```
It's not `to_destroy`, but `to_hide`

arthurBricq replied:
I'm sorry about all these comments about naming, but it is really important when working in a team.

arthurBricq commented on src/world.rs:
```diff
@@ -225,19 +265,30 @@ impl World {
             // Toggle this position
             if let Some(cube_to_toggle) = self.cube_at_mut(pos) {
+
+                // When adding a cube, its neighbors might become non visible
+                // A cube become non visible if it have 5 neighbors before adding a cube
+                // If we test for non visible cube after adding the cube we have extra cube
+                // that were already non visible before
+                if cube_to_toggle.five_neighbors() {
+                    to_destroy.push(cube_to_toggle.position().clone())
+                }
```
Let's talk about this. It is not necessary to add a function `has_five_neighbors` to check if a cube has a 5 neighbors right before incrementing it to 6... It adds a layer of complexity. 

Instead, you can just move this block right after the line below ` cube_to_toggle.add_neighhor();` and then you can use `is_visible` 

Therefore you can remove this function `five_neighbors`. This will be better, because it reduces the knowledge of `World` on if a `Cube` is visible...

arthurBricq reviewed (changes requested):
Overall great job !!! 

There is still some work. I found a few improvements that can be done.

Please be more careful about code style. You have quite a lot of typo, and things like `1.,2.,3.`or `a==b`. These should be `1., 2., 3.` and `a == b`. You can use a formatter to help you with that.

arnaudlorin committed:
Change some variables' name
Remove five_neighbors
Remove apply_action_no_render
Added Option for cubes_to_draw

arnaudlorin committed:
Remove five_neighbors for real

arthurBricq reviewed (approved):
All good to me !

arthurBricq commented:
Fixes https://github.com/arthurBricq/crafty/issues/11
